lib/util/usedPropTypes.js

```diff
@@ -0,0 +1,526 @@
+/**
+ * @fileoverview Common used propTypes detection functionality.
+ */
+'use strict';
+
+const astUtil = require('./ast');
+const versionUtil = require('./version');
+
+// ------------------------------------------------------------------------------
+// Constants
+// ------------------------------------------------------------------------------
+
+const DIRECT_PROPS_REGEX = /^props\s*(\.|\[)/;
+const DIRECT_NEXT_PROPS_REGEX = /^nextProps\s*(\.|\[)/;
+const DIRECT_PREV_PROPS_REGEX = /^prevProps\s*(\.|\[)/;
+const LIFE_CYCLE_METHODS = ['componentWillReceiveProps', 'shouldComponentUpdate', 'componentWillUpdate', 'componentDidUpdate'];
+const ASYNC_SAFE_LIFE_CYCLE_METHODS = ['getDerivedStateFromProps', 'getSnapshotBeforeUpdate', 'UNSAFE_componentWillReceiveProps', 'UNSAFE_componentWillUpdate'];
+
+/**
+ * Checks if a prop init name matches common naming patterns
+ * @param {ASTNode} node The AST node being checked.
+ * @returns {Boolean} True if the prop name matches
+ */
+function isPropAttributeName (node) {
+  return (
+    node.init.name === 'props' ||
+    node.init.name === 'nextProps' ||
+    node.init.name === 'prevProps'
+  );
+}
+
+/**
+ * Checks if the component must be validated
+ * @param {Object} component The component to process
+ * @returns {Boolean} True if the component must be validated, false if not.
+ */
+function mustBeValidated(component) {
+  return Boolean(
+    component &&
+    !component.ignorePropsValidation
+  );
+}
+
+module.exports = function usedPropTypesInstructions(context, components, utils) {
+  const sourceCode = context.getSourceCode();
+  const checkAsyncSafeLifeCycles = versionUtil.testReactVersion(context, '16.3.0');
+
+  /**
+   * Check if we are in a class constructor
+   * @return {boolean} true if we are in a class constructor, false if not
+   */
+  function inComponentWillReceiveProps() {
+    let scope = context.getScope();
+    while (scope) {
+      if (
+        scope.block && scope.block.parent &&
```

Comment:
if we cant get the whole condition on one line, let's put each condition on its own line